src/buildstream/sourcemirror.py

```diff
@@ -0,0 +1,166 @@
+
+
+class SourceMirrorError(BstError):
+    """This exception should be raised by :class:`.SourceMirror` implementations
+    to report errors to the user.
+
+    Args:
+       message: The breif error description to report to the user
+       detail: A possibly multiline, more detailed error message
+       reason: An optional machine readable reason string, used for test cases
+
+    *Since: 2.2*
+    """
+
+    def __init__(
+        self, message: str, *, detail: Optional[str] = None, reason: Optional[str] = None, temporary: bool = False
+    ):
+        super().__init__(message, detail=detail, domain=ErrorDomain.SOURCE, reason=reason)
+
+
+class SourceMirror(Plugin):
+    """SourceMirror()
+
+    Base SourceMirror class.
+
+    All SourceMirror plugins derive from this class, this interface defines how
+    the core will be interacting with SourceMirror plugins.
+
+    *Since: 2.2*
+    """
+
+    # The SourceMirror plugin type is only supported since BuildStream 2.2
+    BST_MIN_VERSION = "2.2"
+
+    COMMON_CONFIG_KEYS = ["name", "kind"]
+    """Common source config keys
+
+    SourceMirror config keys that must not be accessed in configure(), and
+    should be checked for using node.validate_keys().
+    """
+
+    def __init__(
+        self,
+        context: "Context",
+        project: "Project",
+        node: MappingNode,
+    ):
+        # Note: the MappingNode passed here is already expanded with
+        #       the project level base variables, so there is no need
+        #       to expand them redundantly here.
+        #
+
+        # Do local base class parsing first
+        name: str = node.get_str("name")
+
+        # Chain up to Plugin
+        super().__init__(name, context, project, node, "source-mirror")
+
+        self.__aliases: Set[str] = set()
+        # Plugin specific parsing
+        self._configure(node)
+
+    ##########################################################
+    #                      Internal API                      #
+    ##########################################################
+    def _get_alias_uris(self, alias: str) -> List:
+        assert self.__aliases is not None, "Didn't set aliases during configuring time"
+        if alias in self.__aliases:
+            return [self]
+        return []
+
+    ##########################################################
+    #                        Public API                      #
+    ##########################################################
+
+    def set_supported_aliases(self, aliases: List[str]):
+        """Set the aliases for which `self` can translate urls."""
+        assert self._get_configuring(), "Trying to set aliases after configure time"
+        self.__aliases.update(aliases)
+
+    def translate_url(
+        self,
+        *,
+        project_name: str,
+        alias: str,
+        alias_url: str,
+        alias_substitute_url: Optional[str],
+        source_url: str,
+        extra_data: Optional[Dict[str, Any]],
```

Comment:
You no longer intend to change this method signature as per the discussion on the mailing list?

Reply:
Python :shrug: 

I did indeed change where it is used, but forgot it here (and noticed it when trying to address Tristan's comments).